src/llmq/quorums.cpp

```diff
@@ -0,0 +1,406 @@
+// Copyright (c) 2018 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "quorums.h"
+#include "quorums_blockprocessor.h"
+#include "quorums_commitment.h"
+#include "quorums_dkgsession.h"
+#include "quorums_dkgsessionmgr.h"
+#include "quorums_init.h"
+#include "quorums_utils.h"
+
+#include "evo/specialtx.h"
+
+#include "activemasternode.h"
+#include "chainparams.h"
+#include "init.h"
+#include "univalue.h"
+#include "validation.h"
+
+#include "cxxtimer.hpp"
+
+namespace llmq
+{
+
+static const std::string DB_QUORUM_SK_SHARE = "q_Qsk";
+static const std::string DB_QUORUM_QUORUM_VVEC = "q_Qqvvec";
+
+CQuorumManager* quorumManager;
+
+static uint256 MakeQuorumKey(const CQuorum& q)
+{
+    CHashWriter hw(SER_NETWORK, 0);
+    hw << (uint8_t)q.params.type;
+    hw << q.quorumHash;
+    for (const auto& dmn : q.members) {
+        hw << dmn->proTxHash;
+    }
+    return hw.GetHash();
+}
+
+CQuorum::~CQuorum()
+{
+    // most likely the thread is already done
+    stopCachePopulatorThread = true;
+    if (cachePopulatorThread.joinable()) {
+        cachePopulatorThread.join();
+    }
+}
+
+void CQuorum::Init(const uint256& _quorumHash, int _height, const std::vector<CDeterministicMNCPtr>& _members, const std::vector<bool>& _validMembers, const CBLSPublicKey& _quorumPublicKey)
+{
+    quorumHash = _quorumHash;
+    height = _height;
+    members = _members;
+    validMembers = _validMembers;
+    quorumPublicKey = _quorumPublicKey;
+}
+
+bool CQuorum::IsMember(const uint256& proTxHash) const
+{
+    for (auto& dmn : members) {
+        if (dmn->proTxHash == proTxHash) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool CQuorum::IsValidMember(const uint256& proTxHash) const
+{
+    for (size_t i = 0; i < members.size(); i++) {
+        if (members[i]->proTxHash == proTxHash) {
+            return validMembers[i];
+        }
+    }
+    return false;
+}
+
+CBLSPublicKey CQuorum::GetPubKeyShare(size_t memberIdx) const
+{
+    if (quorumVvec == nullptr || memberIdx >= members.size() || !validMembers[memberIdx]) {
+        return CBLSPublicKey();
+    }
+    auto& m = members[memberIdx];
+    return blsCache.BuildPubKeyShare(m->proTxHash, quorumVvec, CBLSId::FromHash(m->proTxHash));
+}
+
+CBLSSecretKey CQuorum::GetSkShare() const
+{
+    return skShare;
+}
+
+int CQuorum::GetMemberIndex(const uint256& proTxHash) const
+{
+    for (size_t i = 0; i < members.size(); i++) {
+        if (members[i]->proTxHash == proTxHash) {
+            return (int)i;
+        }
+    }
+    return -1;
+}
+
+bool CQuorum::WriteContributions(CEvoDB& evoDb)
+{
+    uint256 dbKey = MakeQuorumKey(*this);
+
+    if (quorumVvec != nullptr) {
+        evoDb.GetRawDB().Write(std::make_pair(DB_QUORUM_QUORUM_VVEC, dbKey), *quorumVvec);
+    }
+    if (skShare.IsValid()) {
```

Comment:
Is it possible to have valid `skShare` without having `quorumVvec`? Also, shouldn't we fail if `quorumVvec == nullptr`? 
I.e. maybe
```
bool CQuorum::WriteContributions(CEvoDB& evoDb)
{
    uint256 dbKey = MakeQuorumKey(*this);

    if (quorumVvec == nullptr) return false;

    evoDb.GetRawDB().Write(std::make_pair(DB_QUORUM_QUORUM_VVEC, dbKey), *quorumVvec);

    if (skShare.IsValid()) {
        evoDb.GetRawDB().Write(std::make_pair(DB_QUORUM_SK_SHARE, dbKey), skShare);
    }

    return true;
}
```
?

Reply:
No, without a `quorumVvec` there can't be a `skShare`. Maybe I should completely remove the return value of this method as it doesn't really fail when it doesn't write anything. Also, the return value of this method is actually never used.

Reply:
Yeah, returning `bool` for `Write...` methods is probably not that useful.

Reply:
Changed it to use `void`. Need a re-ACK now